### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by linhpd-axonivy**

### CODE REVIEW

**Comment by weissreto** [comment]
Hi @linhpd-axonivy 
The error is because in the variables.yaml the connection to dockerd is by default defined as Unix Pipe. This will work only on Linux systems. If you want to test it on Windows you need to change to a Named Windows Pipe or TCP/IP. 
And of course you need to install Docker or configure a connection to a remote host that is running Docker.

**Comment by linhpd-axonivy** [comment]
Hi @weissreto 
I installed docker in my computer and set url in variable.yml to "localhost", then I run "dockerDemo/start.ivp" and encounterd this error
image
image
[log.txt](https://github.com/axonivy-market/docker-connector/files/12764049/log.txt)

**Comment by weissreto** [comment]
Hi @linhpd-axonivy
Sorry, I was in holidays the last two weeks. 
Have a look at the comment above URL value in the yaml. There you see how you have to set the URL value for Windows.
Regards
Reto

**Comment by linhpd-axonivy** [comment]
Hi @weissreto 
I just tested the URL value 'localhost' according to your note in 'variables.yaml'. However, I don't know how to get my local Docker URL. I think you should create a guideline in the README that instructs users on how to get the link to configure it. This would be helpful for me and other users to test and use the connector.

**Comment by weissreto** [comment]
If the comment/doc is not clear how should I improve it so that you better understand what you need to configure?

**Comment by linhpd-axonivy** [comment]
Hi @weissreto I don't know if the comments are unclear or if this is my personal issue. As a developer, I use Docker via Docker Desktop, Rancher Desktop, or Podman Desktop to manage my containers because they are easy to install and have a UI. So I don't know how to get the 'URL to communicate with the docker daemon' on my computer. Could you give me a keyword or a guideline to help me find it?

**Comment by linhpd-axonivy** [comment]
I change URL in variable.yml to npipe:////./pipe/dockerDesktopLinuxEngine and start `dockerDemo/start.ivp` and got this error
[log_20_10_2023.txt](https://github.com/axonivy-market/docker-connector/files/13053033/log_20_10_2023.txt)

**Comment by weissreto** [comment]
Let us do a MSTeams Remote Session on Monday. So I can analyze the problem and help you setup your environment correctly.
What is your mail address? Ok, for you at 10:30 CET?
Regards
Reto

**Comment by linhpd-axonivy** [comment]
Hi @weissreto , It's okay for me. This is my email address: `linh.phamduy@axonactive.com`

**Comment by linhpd-axonivy** [comment]
Hi @weissreto, please help me add these line to variables.yml because I don't have push permission
# Window Pipe use Podman (https://podman.io/)
# npipe:////./pipe/docker_engine

File: docker-connector/src/com/axonivy/connector/docker/IvyDockerClient.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+  public static IvyDockerClient instance() {
+    return INSTANCE;
+  }
+
+  private static DockerClient createDockerClient() {
+    var dockerUrl = Ivy.var().get("Docker.URL");
+    var config = DefaultDockerClientConfig.createDefaultConfigBuilder()
+            .withDockerHost(dockerUrl)
+//            .withDockerTlsVerify(true)
```
Comment (linhpd-axonivy) [inline docker-connector/src/com/axonivy/connector/docker/IvyDockerClient.java:37]: Please remove this comment if it is not necessary, or leave a comment explaining why we need it.

File: docker-connector/src/com/axonivy/connector/docker/IvyDockerClient.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    return client.createContainerCmd(image);
+  }
+
+  public ListContainersCmd listContainersCmd() {
+    return client.listContainersCmd();
+  }
+
+  public void stopContainer(String id) {
+    client.stopContainerCmd(id).exec();
+    Ivy.log().info("Container "+id+" stopped");
```
Comment (linhpd-axonivy) [inline docker-connector/src/com/axonivy/connector/docker/IvyDockerClient.java:69]: Please keep the logs if they are needed for debugging or anything else. If not, please remove them.

**Comment by linhpd-axonivy** [comment]
Hi @weissreto, I got error when stop container both case demo and test
[log_23_10_2023.txt](https://github.com/axonivy-market/docker-connector/files/13068804/log_23_10_2023.txt)

File: docker-connector/src/com/axonivy/connector/docker/IvyDockerClient.java (hunk 2 above)
Comment (weissreto) [reply docker-connector/src/com/axonivy/connector/docker/IvyDockerClient.java:69]: I like the logs. Will keep them

**Comment by weissreto** [comment]
The error is strange. Seems like it tries to connect using URL tcp://localhost:2375 instead of the pipe we have configured. Can you update your git repo to the latest version I just pushed and then restart the Designer and try again with the URL for podman?

**Comment by linhpd-axonivy** [comment]
HI @weissreto, I pulled latest version. The demo works well but when I run the tests, it causes the same error as mentioned above.

File: docker-connector-product/README.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
+# Docker Connector
```
Comment (linhpd-axonivy) [inline docker-connector-product/README.md:1]: Hi @andreasbalsiger, 
Could you please help review if the readme file fits the market landscape?
Regards
Linh

**Comment by weissreto** [comment]
> HI @weissreto, I pulled latest version. The demo works well but when I run the tests, it causes the same error as mentioned above.

Hi, I just tried it out on my machine. It works. Also in the Git Hub Actions it works. Let's do another MS teams meeting tomorrow morning CET to find out why it is not running on your machine. 09:30 CET ok for you?

**Comment by linhpd-axonivy** [comment]
Hi, it's ok for me

**Comment by linhpd-axonivy** [comment]
Hi @weissreto , this is the latest logs file 
[log25.txt](https://github.com/axonivy-market/docker-connector/files/13162880/log25.txt)

**Comment by weissreto** [comment]
Hi @linhpd-axonivy 

I've setup a new dev environment with PodMan and was able to reproduce the error you had. 
I've made the code now more robust. It works now with PodMan on my dev environment. 
Can you test if it runs on your machine now too?

Thanks

**Pull request merged**